src/Interpreters/Cache/FileCache.cpp

```diff
@@ -395,640 +339,480 @@ void FileCache::fillHolesWithEmptyFileSegments(
-        if (!cell->releasable())
-        {
-            some_cells_were_skipped = true;
-            continue;
-        }
+        std::vector<FileSegmentMetadata *> remove_cells;
+        remove_cells.reserve(offsets.size());
+        for (auto & [offset, cell] : offsets)
+            remove_cells.push_back(&cell);
 
-        auto file_segment = cell->file_segment;
-        if (file_segment)
+        for (auto & cell : remove_cells)
         {
-            std::unique_lock<std::mutex> segment_lock(file_segment->mutex);
-            file_segment->detach(cache_lock, segment_lock);
-            remove(file_segment->key(), file_segment->offset(), cache_lock, segment_lock);
-        }
-    }
+            /// In ordinary case we remove data from cache when it's not used by anyone.
+            /// But if we have multiple replicated zero-copy tables on the same server
+            /// it became possible to start removing something from cache when it is used
+            /// by other "zero-copy" tables. That is why it's not an error.
+            if (!cell->releasable())
+                continue;
 
-    if (!some_cells_were_skipped)
-    {
-        files.erase(key);
-        removeKeyDirectoryIfExists(key, cache_lock);
+            locked_key->remove(cell->file_segment, lock);
+        }
     }
 }
 
-void FileCache::removeIfReleasable()
+void FileCache::removeAllReleasable()
 {
+    assertInitialized();
+
+    using QueueEntry = IFileCachePriority::Entry;
+    using IterationResult = IFileCachePriority::IterationResult;
+
     /// Try remove all cached files by cache_base_path.
     /// Only releasable file segments are evicted.
     /// `remove_persistent_files` defines whether non-evictable by some criteria files
     /// (they do not comply with the cache eviction policy) should also be removed.
 
-    std::lock_guard cache_lock(mutex);
-
-    std::vector<FileSegmentPtr> to_remove;
-    for (auto it = main_priority->getLowestPriorityReadIterator(cache_lock); it->valid(); it->next())
+    auto lock = cache_guard.lock();
+    LockedCachePriority(lock, *main_priority).iterate([&](const QueueEntry & entry) -> IterationResult
     {
-        const auto & key = it->key();
-        auto offset = it->offset();
-
-        auto * cell = getCell(key, offset, cache_lock);
-        if (!cell)
-        {
-            throw Exception(
-                ErrorCodes::LOGICAL_ERROR,
-                "Cache is in inconsistent state: LRU queue contains entries with no cache cell");
-        }
+        auto locked_key = entry.createLockedKey();
+        auto * cell = locked_key->getKeyMetadata().getByOffset(entry.offset);
 
         if (cell->releasable())
         {
-            auto file_segment = cell->file_segment;
-
-            if (file_segment)
-            {
-                to_remove.emplace_back(file_segment);
-            }
+            cell->queue_iterator = {};
+            locked_key->remove(cell->file_segment, lock);
```

Comment:
Hm, no.
We must clean up key directory strictly under key lock, because yes, here we known that we want to clean up all key directory, but if we release the key lock - new files might be added to cache and we cannot safely clean it up.